### PR TITLE
 xclRegRead and xclRegWrite support 

### PR DESCRIPTION
**Pull request by maxzhen**
This PR introduces two new XRT APIs for accessing register space of a specific CU. Before access, application should first exclusively reserve the target CU.

### CODE REVIEW

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 1]
```diff
@@ -1334,6 +1350,49 @@ uint shim::xclGetNumLiveProcesses()
     return 0;
 }
 
+int shim::xclRegRW(bool rd, uint32_t cu_index, uint32_t offset, uint32_t *datap)
+{
+    std::lock_guard<std::mutex> l(mCuMapLock);
+
+    if (cu_index >= mCuMaps.size()) {
+        std::cerr << "ERROR: xclRegRW: invalid CU index: " <<
```
Comment (stsoe) [inline src/runtime_src/core/pcie/linux/shim.cpp]: Can we use xrt_core::message::send for all cerr calls?

**Comment by maxzhen** [comment]
> 
> 
> This looks good. Much safer than current xclRead/Write, but in reality just a side way to control CUs outside the official execution model.

This is mainly for providing a way to get access to non-control registers, such as read customized IP status...

File: src/runtime_src/core/pcie/linux/shim.cpp (hunk 1 above)
Comment (maxzhen) [reply src/runtime_src/core/pcie/linux/shim.cpp]: > 
> 
> Can we use xrt_core::message::send for all cerr calls?

Will do!

File: src/runtime_src/core/pcie/linux/shim.cpp [hunk 2]
```diff
@@ -1334,6 +1356,55 @@ uint shim::xclGetNumLiveProcesses()
...
+{
+    std::lock_guard<std::mutex> l(mCuMapLock);
+
+    if (cu_index >= mCuMaps.size()) {
+        std::string err = "xclRegRW: invalid CU index: ";
+        err += std::to_string(cu_index);
+        errlog(err);
+        return -EINVAL;
+    }
+    if (offset >= mCuMapSize || (offset % sizeof(uint32_t)) != 0) {
```
Comment (sonals) [inline src/runtime_src/core/pcie/linux/shim.cpp]: offset & 0x3 to determine if it is multiple of 4?

**Pull request merged**